text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,153 @@
+
+### Backend Token Hygiene
+
+API tokens should be stored in the database with a cryptographically secure hash (minimum SHA-256). Salting is not necessary since these values are sufficient in length.
+
+An additional column identifying the API token version will be added in order to correctly identify legacy API tokens.
+
+### Secret Verification Service
+
+In order to integrate with Github's Secret Scanning service, we will need to have an available service for them to submit potential matches to.
+
+More details about Github's secret scanning program can be found [here](https://docs.github.com/en/developers/overview/secret-scanning-partner-program).
+
+At a high-level, we will need to be able to:
+
+- Receive a list of one or more potential tokens
+- [Verify the signature of the request](https://docs.github.com/en/developers/overview/secret-scanning-partner-program#implement-signature-verification-in-your-secret-alert-service)
+- Determine if the token is a true positive or a false positive
+  - send this feedback back to Github
+- Disable/revoke the compromised token
+- Send a notification to the appropriate user
+
+> :memo: This secret verification service is not an immediate need or a dependency of implementing the improved API tokens.
+
+# Background
+
+Accidental credential leaks [happen](https://github.com/ChALkeR/notes/blob/master/Do-not-underestimate-credentials-leaks.md). Even though we provide several tools for users to limit the storage of sensitive information in Sentry it can still happen.
+
+Our support for various authentication methods (including 2FA, SSO, and SAML) helps mitigate access of potentially sensitive information, but API tokens cannot support these additional authentication gates.
+
+We can help customers further protect their accounts and data by providing a means of auto-detecting leaked API tokens.
+
+# Options Considered
+
+## Option #1
+
+This option strives to limit impact to the database by avoiding large bulk operations. To do this we'll distribute the complete implementation over two features. This also allows for a smooth transition of self-hosted instances keeping pace with releases.
+
+**In the first release:**
+
+1. The frontend is updated to no longer display the token value for existing tokens.
+2. The frontend is updated to accept a `name` when creating new tokens.
+3. A notification/warning in the UI should be displayed recommending users recreate their tokens, resulting in the new token version.
+4. New tokens are only displayed once at creation time to the user.
+5. New tokens are created following the new format.
+   - A Django migration will be needed to add fields: `hashed_token`, `version`, and `name`.
+   - legacy tokens will be `version = 1` and new tokens `version = 2`
+6. As old tokens are used, they are hashed and stored in the database.
+   - This does not update the token to the new format.
+   - The plaintext token value should be removed in this action.
+
+**In the second release:**
```

Comment:
I would have expected two more releases afterward to 1) remove the special case frontend code 2) remove the API compatibility code.

Otherwise these become a permanent technical burden. It may be small but many such small burdens add up over time.

Reply:
See https://github.com/getsentry/rfcs/pull/32#discussion_r1010581248. I'll update the document to make this more clear.